src/FNA3D_Driver_Vulkan.c

```diff
@@ -6490,256 +6526,345 @@ static void VULKAN_INTERNAL_FlushCommandsAndPresent(
+	swapchainData->swapchainSwizzle.g = VK_COMPONENT_SWIZZLE_IDENTITY;
+	swapchainData->swapchainSwizzle.b = VK_COMPONENT_SWIZZLE_IDENTITY;
+	swapchainData->swapchainSwizzle.a = VK_COMPONENT_SWIZZLE_IDENTITY;
+
 	if (!VULKAN_INTERNAL_ChooseSwapSurfaceFormat(
-		renderer->swapchainFormat,
-		swapChainSupportDetails.formats,
-		swapChainSupportDetails.formatsLength,
-		&surfaceFormat
+		swapchainData->swapchainFormat,
+		swapchainSupportDetails.formats,
+		swapchainSupportDetails.formatsLength,
+		&swapchainData->surfaceFormat
 	)) {
 		FNA3D_LogWarn("RGBA8 swapchain unsupported, falling back to BGRA8 with swizzle");
-		renderer->swapchainFormat = renderer->backBufferIsSRGB
+		swapchainData->swapchainFormat = renderer->backBufferIsSRGB
 			? VK_FORMAT_B8G8R8A8_SRGB
 			: VK_FORMAT_B8G8R8A8_UNORM;
-		renderer->swapchainSwizzle.r = VK_COMPONENT_SWIZZLE_B;
-		renderer->swapchainSwizzle.g = VK_COMPONENT_SWIZZLE_G;
-		renderer->swapchainSwizzle.b = VK_COMPONENT_SWIZZLE_R;
-		renderer->swapchainSwizzle.a = VK_COMPONENT_SWIZZLE_A;
+		swapchainData->swapchainSwizzle.r = VK_COMPONENT_SWIZZLE_B;
+		swapchainData->swapchainSwizzle.g = VK_COMPONENT_SWIZZLE_G;
+		swapchainData->swapchainSwizzle.b = VK_COMPONENT_SWIZZLE_R;
+		swapchainData->swapchainSwizzle.a = VK_COMPONENT_SWIZZLE_A;
+
 		if (!VULKAN_INTERNAL_ChooseSwapSurfaceFormat(
-			renderer->swapchainFormat,
-			swapChainSupportDetails.formats,
-			swapChainSupportDetails.formatsLength,
-			&surfaceFormat
+			swapchainData->swapchainFormat,
+			swapchainSupportDetails.formats,
+			swapchainSupportDetails.formatsLength,
+			&swapchainData->surfaceFormat
 		)) {
-			SDL_free(swapChainSupportDetails.formats);
-			SDL_free(swapChainSupportDetails.presentModes);
 			FNA3D_LogError("Device does not support swap chain format");
+			if (swapchainSupportDetails.formatsLength > 0)
+			{
+				SDL_free(swapchainSupportDetails.formats);
+			}
+			if (swapchainSupportDetails.presentModesLength > 0)
+			{
+				SDL_free(swapchainSupportDetails.presentModes);
+			}
 			return CREATE_SWAPCHAIN_FAIL;
 		}
 	}
 
 	if (!VULKAN_INTERNAL_ChooseSwapPresentMode(
 		renderer->presentInterval,
-		swapChainSupportDetails.presentModes,
-		swapChainSupportDetails.presentModesLength,
-		&presentMode
+		swapchainSupportDetails.presentModes,
+		swapchainSupportDetails.presentModesLength,
+		&swapchainData->presentMode
 	)) {
-		SDL_free(swapChainSupportDetails.formats);
-		SDL_free(swapChainSupportDetails.presentModes);
 		FNA3D_LogError("Device does not support swap chain present mode");
+		if (swapchainSupportDetails.formatsLength > 0)
+		{
+			SDL_free(swapchainSupportDetails.formats);
+		}
+		if (swapchainSupportDetails.presentModesLength > 0)
+		{
+			SDL_free(swapchainSupportDetails.presentModes);
+		}
 		return CREATE_SWAPCHAIN_FAIL;
 	}
 
-	extent = ChooseSwapExtent(
-		renderer->deviceWindowHandle,
-		swapChainSupportDetails.capabilities
+	/* TODO: validate swapchain extents */
```

Comment:
Remaining TODO. In particular, the 0,0 extent is actually really important for surface suspension on Windows and Wayland, so we'll need to reincorporate this where it's been removed.

Other than that this is all fine.